### PR TITLE
chore(ci): add script to generate MSR test pipelines

### PR DESCRIPTION
**Pull request by sudanl0**
rdmsr and wrmsr/cpuid snapshot restore tests have
` "number_of_instances * number_of_supported_kernels * number_of_tests"`
buildkite jobs and if we need to make a common change we will have to modify each of these buildkite jobs manually which is not optimal.

This script aims to centralise the common task so that common changes can be minimised and done only at one place in the script. 

## Changes

Create a script to generate pipelines for rdmsr test, wrmsr snapshot/restore test and cpuid snapshot/restore test.

## Reason

To avoid manually updating buildkite jobs when a common change is needed in the tests.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [ ] If a specific issue led to this PR, this PR closes the issue.
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] API changes follow the [Runbook for Firecracker API changes][2].
- [ ] User-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.
- [ ] New `TODO`s link to an issue.
- [ ] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [ ] This functionality can be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: .buildkite/common.py [hunk 1]
```diff
@@ -14,6 +16,94 @@
...
+    # get a copy to edit and cleanup the data
+    step_copy = copy.deepcopy(test_step)
+
+    # add the instance and kernel version to the step
+    step_copy["agents"] = [f"instance={instance}", f"kv={kv}"]
+
+    if LABEL in step_copy.keys():
+        step_copy[LABEL] = step_copy[LABEL] + label
+    else:
+        step_copy[LABEL] = label
```
Comment (pb8o) [inline .buildkite/common.py:52]: Nit:

```python
step_copy.setdefault(LABEL, "")
step_copy[LABEL] += label
```

File: .buildkite/common.py [hunk 2]
```diff
@@ -14,6 +16,94 @@
...
+
+    if EMOJI in step_copy.keys():
+        step_copy[LABEL] = step_copy[EMOJI] + step_copy[LABEL]
+        # cleanup keys that are invalid as per buildkite step
+        del step_copy[EMOJI]
+
+    for i, _ in enumerate(step_copy[COMMAND]):
+        # this is so that artifacts can be download from a unique step
+        if "--step" in step_copy[COMMAND][i]:
+            step_copy[COMMAND][i] = step_copy[COMMAND][i].format(label)
```
Comment (pb8o) [inline .buildkite/common.py:66]: This and `key` above are just working around what could be a simple fix in the tests

**Review by pb8o — CHANGES_REQUESTED** [review]

File: .buildkite/common.py [hunk 3]
```diff
@@ -14,6 +16,94 @@
 
 DEFAULT_KERNELS = ["linux_4.14", "linux_5.10"]
 
+# Define frequently used keywords having double quotes to avoid
```
[the remark this answers is not in the document]
Comment (sudanl0) [reply .buildkite/common.py:19]: It would be helpful to have them in common file for new tests which need the "wait" step.

File: .buildkite/pipeline_cpu_template.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        {
+            COMMAND  : [
+                "tools/devtool -y test -- \
+                 -s --nonci integration_tests/functional/test_cpu_features.py \
+                 -k 'test_cpu_wrmsr_snapshot or test_cpu_cpuid_snapshot'"
+            ],
+            EMOJI    : "📸",
+            # lebel here is important since it will be used by
+            # restore test to find the uploaded artifacts
+            LABEL    : "wrmsr_cpuid_snapshot",
```
[the remark this answers is not in the document]
Comment (sudanl0) [reply .buildkite/pipeline_cpu_template.py]: we have use cases like snapshot on m5d.metal and restore on m6i.metal so adding the instance name won't be easy. 
And lets say I am creating a snapshot on m5d.metal and restoring it on m6i.metal, I can try to upload the artifacts to
tests/snapshot_artifacts/m5d.metal_to_m6i.metal/**/*
but if there is some other user running the same test in parallel then this parallel test will also upload the data to the same path and the path can never be unique.
so I think the "key" and "--step" approach is more reliable.

File: .buildkite/common.py (hunk 1 above)
Comment (sudanl0) [reply .buildkite/common.py]: thanks!

File: .buildkite/common.py (hunk 2 above)
Comment (sudanl0) [reply .buildkite/common.py]: added some explanation here https://github.com/firecracker-microvm/firecracker/pull/3458#discussion_r1114530928

File: .buildkite/common.py [hunk 5]
```diff
@@ -32,11 +33,15 @@ def group(label, command, instances, platforms, agent_tags=None, **kwargs):
     # Use the 1st character of the group name (should be an emoji)
     label1 = label[0]
     steps = []
+    if isinstance(commands, str):
+        commands = [commands]
```
Comment (pb8o) [inline .buildkite/common.py:37]: I thought this was already in main? Maybe an artifact of a rebase...

File: .buildkite/pipeline_cpu_template.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+import json
+from enum import Enum
+
+from common import group
+
+class bkstep(str, Enum):
+    LABEL="label"
+    TIMEOUT="timeout"
+    COMMAND="commands"
+    ARTIFACTS="artifact_paths"
```
Comment (pb8o) [inline .buildkite/pipeline_cpu_template.py:17]: To reformat:
```
cd tests
black ../.buildkite/pipeline_cpu_template.py
isort ../.buildkite/pipeline_cpu_template.py
```

File: .buildkite/pipeline_cpu_template.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        bkstep.LABEL: "📖 rdmsr",
+        "instances": ["m5d.metal", "m6a.metal", "m6i.metal"],
+    },
+    "cpuid_wrmsr": {
+        "snapshot": {
+            bkstep.COMMAND: [
+                "tools/devtool -y test -- \
+                 -s -ra --nonci  --log-cli-level=INFO \
+                 integration_tests/functional/test_cpu_features.py \
+                 -k 'test_cpu_wrmsr_snapshot or test_cpu_cpuid_snapshot'",
```
Comment (pb8o) [inline .buildkite/pipeline_cpu_template.py:36]: nit: I would keepit in a single line and let black reformat it as it wants.

File: .buildkite/pipeline_cpu_template.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+            ((("al2", "linux_4.14"), "m5d.metal"), (("al2", "linux_4.14"), "m6i.metal")),
+            ((("al2", "linux_4.14"), "m6i.metal"), (("al2", "linux_4.14"), "m5d.metal")),
+            ((("al2", "linux_4.14"), "m6i.metal"), (("al2", "linux_4.14"), "m6i.metal")),
+            ((("al2", "linux_4.14"), "m6a.metal"), (("al2", "linux_4.14"), "m6a.metal")),
+            ((("al2", "linux_5.10"), "m5d.metal"), (("al2", "linux_5.10"), "m5d.metal")),
+            ((("al2", "linux_5.10"), "m5d.metal"), (("al2", "linux_5.10"), "m6i.metal")),
+            ((("al2", "linux_5.10"), "m6i.metal"), (("al2", "linux_5.10"), "m5d.metal")),
+            ((("al2", "linux_5.10"), "m6i.metal"), (("al2", "linux_5.10"), "m6i.metal")),
+            ((("al2", "linux_5.10"), "m6a.metal"), (("al2", "linux_5.10"), "m6a.metal")),
+        ],
```
Comment (pb8o) [inline .buildkite/pipeline_cpu_template.py:67]: Consider generating this. Maybe keep the data here and generate in group_snapshot_restore

```python
cross_instances = {
  "m5d": ["m6i"],
  "m6i": ["m5d"],
}
src_instances = ["m5d", "m6i", "m6a"]

combs = []
for dp in DEFAULT_PLATFORMS:
    for src_instance in src_instances:
        for dst_instance in [src_instance] + others.get(cross_instances, []):
            combs.append(((dp, src_instance), (dp, dst_instance)))
```

File: .buildkite/pipeline_cpu_template.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    )
+    test_args = parser.parse_args()
+
+    if test_args.test == "rdmsr":
+        test_group = group_rdmsr(cpu_template_test[test_args.test])
+    else:
+        test_group = group_snapshot_restore(cpu_template_test[test_args.test])
+
+    pipeline = {
+                "agents" : {"queue": "public-prod-us-east-1"},
```
Comment (pb8o) [inline .buildkite/pipeline_cpu_template.py:159]: Better to remove this and put it on the pipeline.

File: .buildkite/pipeline_cpu_template.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+                "steps"  : test_group
+               }
+    print(json.dumps(pipeline, indent=4, sort_keys=True, ensure_ascii=False))
```
Comment (pb8o) [inline .buildkite/pipeline_cpu_template.py:162]: Can use `pipeline_to_json` here.

File: .buildkite/pipeline_cpu_template.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+def group_snapshot_restore(test_step):
+    """
+    Generate a group step with specified parameters for each instance
+    and kernel combination and handle "wait" command between steps
+    https://buildkite.com/docs/pipelines/group-step
+    """
+    groups = []
+
+    snapshot_instances = list(set([instance for (_,instance),(_,_) in test_step["snapshot_restore_combinations"]]))
+    snapshot_platforms = list(set([plat for (plat,_),(_,_) in test_step["snapshot_restore_combinations"]]))
```
Comment (pb8o) [inline .buildkite/pipeline_cpu_template.py:96]: This could be just `DEFAULT_PLATFORMS`

File: .buildkite/pipeline_cpu_template.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+def group_snapshot_restore(test_step):
+    """
+    Generate a group step with specified parameters for each instance
+    and kernel combination and handle "wait" command between steps
+    https://buildkite.com/docs/pipelines/group-step
+    """
+    groups = []
+
+    snapshot_instances = list(set([instance for (_,instance),(_,_) in test_step["snapshot_restore_combinations"]]))
```
Comment (pb8o) [inline .buildkite/pipeline_cpu_template.py:95]: This could `src_instances` from my comment above

File: .buildkite/common.py [hunk 13]
```diff
@@ -6,6 +6,7 @@
 """
 
 import json
+import pdb
```
Comment (pb8o) [inline .buildkite/common.py:9]: debug code?

File: .buildkite/common.py (hunk 5 above)
Comment (sudanl0) [reply .buildkite/common.py:37]: I had created the PR before this was mainlined. I'll do a cleanup of common.py

**Review by pb8o — DISMISSED** [review]

File: .buildkite/pipeline_cpu_template.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+
+def group_rdmsr(tests):
+    """
+    Generate a group step with specified parameters for each instance
+    and kernel combination
+    https://buildkite.com/docs/pipelines/group-step
+    """
+    # pdb.set_trace()
```
Comment (kalyazin) [inline .buildkite/pipeline_cpu_template.py:68]: do we need this still?

File: .buildkite/pipeline_cpu_template.py [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+            "tools/devtool -y test -- -s -ra --nonci --log-cli-level=INFO integration_tests/functional/test_cpu_features.py -k 'test_cpu_rdmsr' "
+        ],
+        BkStep.LABEL: "📖 rdmsr",
+        "instances": ["m5d.metal", "m6a.metal", "m6i.metal"],
+        "platforms": DEFAULT_PLATFORMS,
+    },
+    "cpuid_wrmsr": {
+        "snapshot": {
+            BkStep.COMMAND: [
+                "tools/devtool -y test -- -s -ra --nonci  --log-cli-level=INFO integration_tests/functional/test_cpu_features.py -k 'test_cpu_wrmsr_snapshot or test_cpu_cpuid_snapshot'",
```
Comment (kalyazin) [inline .buildkite/pipeline_cpu_template.py:36]: nit: here an below: double space.

File: .buildkite/pipeline_cpu_template.py [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+            }
+        )
+
+    groups.append({"group": "📸 restores snapshots", "steps": steps})
+    return groups
+
+
+def main():
+    """
+    Generate group template required trigger pipelines for
```
Comment (kalyazin) [inline .buildkite/pipeline_cpu_template.py:142]: nit: `group template required trigger pipelines` -- this is hard to parse for me...

File: .buildkite/pipeline_cpu_template.py (hunk 14 above)
Comment (sudanl0) [reply .buildkite/pipeline_cpu_template.py]: removed.

File: .buildkite/pipeline_cpu_template.py (hunk 15 above)
Comment (sudanl0) [reply .buildkite/pipeline_cpu_template.py]: removed.

File: .buildkite/pipeline_cpu_template.py (hunk 16 above)
Comment (sudanl0) [reply .buildkite/pipeline_cpu_template.py]: "to" was missing

**Review by kalyazin — APPROVED** [review]

**Review by pb8o — APPROVED** [review]
